### ISSUE TITLE
New build stall / unexplained error on xcode 8.3 image

### ISSUE THREAD

**Issue by matthew-brett** [travis-ci/travis-ci#8703]
I have a project that runs builds on various xcode images.

For the last 2 days, the XCode 8.3 image build (only) stalls, and then generates an error, with no useful information in the log:

https://travis-ci.org/matthew-brett/multibuild/builds/297159959

This started for a Cron build at 2017-11-03 15.20 

https://travis-ci.org/matthew-brett/multibuild/builds/296772991

This built exactly the same commit as the previous build, which passed:

https://travis-ci.org/matthew-brett/multibuild/builds/296253660

**Comment by BanzaiMan** [comment]
Sorry for the delay. Is this still a problem for you?

**Comment by matthew-brett** [comment]
Yes, still a problem : https://travis-ci.org/matthew-brett/multibuild/jobs/300597675 .

**Comment by BanzaiMan** [comment]
The likely explanation is that one of the commands you are running on this image is exiting the script prematurely. Try turning on `verbose` option (`set -v`) or `xtrace` (`set -x`). It is curious that this is the only image that exhibits the behavior, though.

**Comment by matthew-brett** [comment]
Set -x test here : https://travis-ci.org/matthew-brett/multibuild/jobs/300597675

**Comment by matthew-brett** [comment]
Oops - sorry - I mean https://travis-ci.org/matthew-brett/multibuild/builds/300109458

**Comment by matthew-brett** [comment]
I believe this is a minimal reproducible example:

```
language: generic
os: osx
osx_image: xcode8.3

script:
  - set -ex
  - shell_session_update() { :; }
  - mkdir tmp_repos
  - cd tmp_repos
```

See: https://github.com/matthew-brett/travis-testing/commit/8f2981d8fb1f870971c5abe0102c8e5955e0d14e and https://travis-ci.org/matthew-brett/travis-testing/builds/300665731

The override for `shell_session_update` is for an earlier bug discussed at https://github.com/direnv/direnv/issues/210

**Comment by takluyver** [comment]
I think I've found a workaround for this:

```shell
unset -f cd
```

rvm [overrides cd with a function](https://github.com/rvm/rvm/blob/ff63fa117bd05e581ecbf2b0378d9ddbb263deb1/scripts/cd#L14) so that it can hook into it to run some scripts from the new working directory. The workaround removes that, so `cd` is the normal shell builtin again. I don't know what the impact is for Ruby projects, but for other languages it should be safe to get rid of the special Ruby magic.

**Comment by maelvalais** [comment]
Same on [our build](https://travis-ci.org/touist/touist/jobs/308892638): just after a ˋcd`, the error

    /Users/travis/.rvm/scripts/functions/support: line 57: shell_session_update: command not found

shows but no explanation whatsoever.
I’ll try your unset workaround asap!

**Comment by jakirkham** [comment]
Same story with [`pushd` and `popd`]( https://github.com/rvm/rvm/blob/ff63fa117bd05e581ecbf2b0378d9ddbb263deb1/scripts/cd#L15-L16 ).

xref: https://github.com/scipy/scipy/issues/8192#issuecomment-348749648
xref: https://github.com/xoviat/scipy/pull/6
ref: https://travis-ci.org/andyfaff/scipy/jobs/310776437

**Comment by sormuras** [comment]
xcode 9.3 is also affected: https://travis-ci.org/sormuras/bach/jobs/384571589

**Comment by illume** [comment]
For people tracking this, it looks like there are at least three open issues concerning shell_session_update:

- https://github.com/travis-ci/travis-ci/issues/9511
- https://github.com/travis-ci/travis-ci/issues/9486
- https://github.com/travis-ci/travis-ci/issues/6522

**Comment by asanakoy** [comment]
I have the same issue. Looks like it's not solved yet. 
Call `cd` yields 
`/Users/travis/.rvm/scripts/functions/support: line 78: shell_session_update: command not found`.
